racket/src/thread/future.rkt

```diff
@@ -0,0 +1,441 @@
+#lang racket/base
+
+(require "check.rkt"
+         "internal-error.rkt"
+         "engine.rkt"
+         "atomic.rkt"
+         "parameter.rkt"
+         "../common/queue.rkt"
+         "thread.rkt")
+
+(provide futures-enabled?
+         processor-count
+         current-future
+         future
+         future?
+         would-be-future
+         touch
+         future-block
+         current-future-prompt
+         future:condition-broadcast
+         future:condition-signal
+         future:condition-wait
+         future:make-condition
+         lock-release
+         lock-acquire
+         make-lock
+         with-lock
+         )
+;; ---------------------------- locks ------------------------------------
+
+(define-syntax-rule (with-lock lock expr ...)
+  (begin
+    (lock-acquire lock)
+    (begin0
+        (let () expr ...)
+      (lock-release lock))))
+
+(struct future-lock* (box owner))
+
+(define (make-lock)
+  (future-lock* (box 0) (box #f)))
+
+;;  cannot use atomics in locking code because it is used by code that
```

Comment:
Is it only because you want to check the owner of a lock that locks have to be in "future.rkt"? If so, consider a lock interface where you pass an owner explicitly, and that way locks don't have to be combined here and implemented in multiple places.